packages/theme/src/cli/utilities/asset-ignore.ts

```diff
@@ -13,10 +13,28 @@ export function applyIgnoreFilters<T extends {key: string}>(
   const ignoreOptions = options.ignore ?? []
   const onlyOptions = options.only ?? []
 
-  return files
-    .filter(filterBy(shopifyIgnore, '.shopifyignore'))
-    .filter(filterBy(ignoreOptions, '--ignore'))
-    .filter(filterBy(onlyOptions, '--only', true))
+  const [normalShopifyPatterns = [], negatedShopifyPatterns = []] = filterRegexValues(shopifyIgnore)
+  const [normalIgnorePatterns = [], negatedIgnorePatterns = []] = filterRegexValues(ignoreOptions)
+  const [normalOnlyPatterns = [], negatedOnlyPatterns = []] = filterRegexValues(onlyOptions)
+
+  let filteredFiles = files.filter(filterBy(normalShopifyPatterns, '.shopifyignore'))
+  filteredFiles = filteredFiles.filter(filterBy(normalIgnorePatterns, '--ignore'))
+  filteredFiles = filteredFiles.filter(filterBy(normalOnlyPatterns, '--only', true))
+
+  if (negatedShopifyPatterns.length > 0) {
+    filteredFiles = filteredFiles.concat(files.filter(filterBy(negatedShopifyPatterns, '.shopifyignore', true)))
+  }
+  if (negatedIgnorePatterns.length > 0) {
+    filteredFiles = filteredFiles.concat(files.filter(filterBy(negatedIgnorePatterns, '--ignore', true)))
+  }
+  if (negatedOnlyPatterns.length > 0) {
+    filteredFiles = filteredFiles.filter(filterBy(negatedOnlyPatterns, '--only'))
+  }
+
+  const uniqueFiles = Array.from(new Set(filteredFiles.map((file) => file.key))).map((key) =>
+    filteredFiles.find((file) => file.key === key),
+  )
+  return uniqueFiles
```

Comment:
Can you try using [uniqBy](https://github.com/Shopify/cli/blob/e58493f44c616e2dae2847086158287f6f777bb0/packages/cli-kit/src/public/common/array.ts#L46-L57) for this grabbing unique files?